### PR TITLE
Add clients_daily query

### PR DESCRIPTION
**Pull request by relud**

### CODE REVIEW

File: udf/udf_require_whole_geo.sql [hunk 1]
```diff
@@ -0,0 +1,13 @@
+CREATE TEMP FUNCTION
+  udf_require_whole_geo(country STRING,
+    city STRING,
+    geo_subdivision1 STRING,
+    geo_subdivision2 STRING) AS ( IF(country != '??',
```
[the remark this answers is not in the document]
Comment (relud) [reply udf/udf_require_whole_geo.sql]: I'll add a comment, it comes from hindsight's geoip decoding

File: sql/clients_daily_v7.sql [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        NULLIF(city,
+          '??') AS city,
+        NULLIF(geo_subdivision1,
+          '??') AS geo_subdivision1,
+        NULLIF(geo_subdivision2,
+          '??') AS geo_subdivision2),
+      NULL));
+  --
+WITH
+  -- normalize client_id and rank by document_id
```
[the remark this answers is not in the document]
Comment (relud) [reply sql/clients_daily_v7.sql]: order by, sorry I wrote this comment a while back, not sure what i was thinking

File: tests/clients_daily_v7/test_first_value_falsey/main_summary_v4.schema.json [hunk 3]
```diff
@@ -0,0 +1,8736 @@
+[
```
[the remark this answers is not in the document]
Comment (relud) [reply tests/clients_daily_v7/test_first_value_falsey/main_summary_v4.schema.json]: yes, we can, for this PR i was just continuing to use what was already in place

[the remark this answers is not in the document]
Comment (jklukas) [reply tests/clients_daily_v7/test_first_value_falsey/main_summary_v4.schema.json]: Feel free to ignore for the scope of this PR.

**Comment by relud** [comment]
> Is this using the auto-generation machinery, or is this hand-built? Is this expected to match exactly with the imported parquet-based clients_daily?

This is hand-built to exactly match the schema of clients_daily_v6 (the imported parquet-based one). Matching the data exactly isn't really possible because v6 uses `FIRST_VALUE` with no ordering and it's very non-trivial to add ordering, so it's effectively `ANY_VALUE`. So we could probably use `udf_mode_last` instead, I just forgot about switching to that when I hit the PR button.

**Comment by jklukas** [comment]
> So we could probably use udf_mode_last instead, I just forgot about switching to that when I hit the PR button

I think we might as well go ahead and do that if we aren't able to match existing clients_daily exactly anyway.

I'm also just realizing that since we're ordering by timestamp DESC, we have probably been doing the wrong thing (breaking ties by taking the earliest seen value rather than latest) by using `mode_last`. Should we be using `udf_mode_first` instead?

File: sql/clients_daily_v7.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (relud) [reply sql/clients_daily_v7.sql]: oh, i see now. `partition by` is definitely better, using that.

**Comment by relud** [comment]
I think `mode_last` is still the right method, and we should consider fixing and backfilling the other tables that coincidentally did `mode_first`

File: udf/udf_null_if_empty_list.sql [hunk 4]
```diff
@@ -0,0 +1,4 @@
+CREATE TEMP FUNCTION
+  udf_null_if_empty_list(list ANY TYPE) AS ( IF(ARRAY_LENGTH(list.list) > 0,
+      list,
+      NULL) );
```
[the remark this answers is not in the document]
Comment (relud) [reply udf/udf_null_if_empty_list.sql]: I will add a comment, it throws this error: `NULLIF is not defined for arguments of type ARRAY<STRUCT<element STRING>>`

**Comment by jklukas** [comment]
> I think mode_last is still the right method, and we should consider fixing and backfilling the other tables that coincidentally did mode_first

Got it. I see now that you're sorting ASC here in both frames, so `mode_last` is correct. I agree that other tables should be fixed to match the pattern used here, which is easier to reason about than sorting DESC.

**Comment by relud** [comment]
I've updated this to use `mode_last`, but it's throwing `BigQuery error in query operation: Resources exceeded during query execution: Not enough resources for query planning - too many subqueries or query is too complex.` so we'll see how tricky i have to get to avoid that

**Comment by relud** [comment]
I was able to do `mode_last` without getting a `too complex` exception by combining all of the `udf_mode_last` UDFs into a single (nested) subquery expression, but [it's not pretty](https://github.com/mozilla/bigquery-etl/blob/clients_daily_mode_last/sql/clients_daily_v7.sql#L322-L447), so I'm going to break that out into a separate PR for further discussion.

**Pull request merged**